### PR TITLE
Add polynomial evaluator

### PR DESCRIPTION
**Pull request by sethrj**
- Multiplied a `1e-3` inside `calc_step_min` into the precalculated data rather than runtime
- Added `PolyEvaluator` class to simplify polynomial evalulation expressions
- Rewrote a couple of expressions using the evaluator (livermore, MSC, LPM)
- Refactored the LPM g/phi calculation to reduce code duplication and have clearer piecewise regimes
- Eliminated a few local variables in MSC that were just "a - 1" or "1 - c" or other trivial arithmetic expressions (arithmetic is way less expensive than registers/local writes on GPU)

### CODE REVIEW

File: src/physics/em/detail/UrbanMscScatter.hh [hunk 1]
```diff
@@ -355,24 +351,24 @@ CELER_FUNCTION real_type UrbanMscScatter::sample_cos_theta(Engine& rng,
             if (BernoulliDistribution(prob)(rng))
             {
                 // Sample \f$ \cos\theta \f$ from \f$ g_1(\cos\theta) \f$
-                result = 1 + std::log(ea + generate_canonical(rng) * eaa) * x;
+                UniformRealDistribution<real_type> sample_inner(ea, 1);
```
Comment (whokion) [inline src/physics/em/detail/UrbanMscScatter.hh:354]: Very good case not to have eea!

**Review by whokion — APPROVED** [review]
Like PolyEvaluator! Should have many other places to apply.

File: src/physics/em/UrbanMscModel.cc [hunk 2]
```diff
@@ -143,26 +146,25 @@ auto UrbanMscModel::calc_material_data(const MaterialView& material_view)
...
+    data.d[1]  = PolyQuad(4.7526e-1, 1.7694, -3.3885e-1)(z13);
+    data.d[2]  = PolyQuad(2.3683e-1, -1.8111, 3.2774e-1)(z13);
+    data.d[3]  = PolyQuad(1.7888e-2, 1.9659e-2, -2.6664e-3)(z13);
 
     data.z23 = ipow<2>(z13);
 
     // Parameters for the step minimum calculation
-    data.stepmin_a = 27.725 / (1 + 0.203 * zeff);
-    data.stepmin_b = 6.152 / (1 + 0.111 * zeff);
+    data.stepmin_a = 1e3 * 27.725 / (1 + 0.203 * zeff);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/UrbanMscModel.cc:166]: I'm happy to put the 1e3 somewhere else if it has a physical meaning: the raw number alone made it seem like another magic constant or magic tolerance so I figured it should be part of the stored data.

[the remark this answers is not in the document]
Comment (whokion) [reply src/physics/em/UrbanMscModel.cc:166]: Either way should be fine as the correct result is same!  Up to your preference.

[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/UrbanMscModel.cc:166]: Ideally we could label the 1e3 as "gev to mev " (or whatever) and process it into the constants themselves. I'd prefer readability over "micro-optimization", but both if possible, and the original form had neither?

[the remark this answers is not in the document]
Comment (whokion) [reply src/physics/em/UrbanMscModel.cc:166]: You are right that the original form has also two fuzzy constants,
`return lambda0*1.e-3/(2.e-3+rat*(msc[idx]->stepmina+msc[idx]->stepminb*rat));`
which does not have clear origins - we all agree to hate this model, may be even macroscopic msc itself, which does not exist in reality.

File: src/physics/em/detail/LPMCalculator.hh [hunk 3]
```diff
@@ -161,90 +163,82 @@ CELER_FUNCTION auto LPMCalculator::operator()(real_type epsilon) -> LPMFunctions
...
-            }
-        }
+        return 1 - real_type(0.01190476) / ipow<4>(s);
     }
+}
 
-    return result;
+//---------------------------------------------------------------------------//
+/*!
+ * Compute the LPM suppression function \f$ \phi(s) \f$.
```
Comment (amandalund) [inline src/physics/em/detail/LPMCalculator.hh:215]:
```suggestion
 * Compute the LPM suppression function \f$ \G(s) \f$.
```

**Review by amandalund — APPROVED** [review]
Very nice!

**Comment by sethrj** [comment]
Whoops, sorry for force-pushing...

**Pull request merged**